src/gc/gcinterface.h

```diff
@@ -0,0 +1,499 @@
+
+    // Gets the duration of the last GC that occured for the given generation.
+    virtual size_t GetLastGCDuration(int generation) = 0;
+
+    // Gets a timestamp for the current moment in time.
+    virtual size_t GetNow() = 0;
+
+    // Returns whether or not a GC is in progress.
+    virtual BOOL IsGCInProgressHelper(BOOL bConsiderGCStart = FALSE) = 0;
+
+    // Returns the number of GCs that have occured. Mainly used for
+    // sanity checks asserting that a GC has not occured.
+    virtual unsigned GetGcCount() = 0;
+
+    // Sets cards after an object has been memmoved. 
+    virtual void SetCardsAfterBulkCopy(Object** obj, size_t length) = 0;
+
+    // Gets whether or not the home heap of this alloc context matches the heap
+    // associated with this thread.
+    virtual bool IsThreadUsingAllocationContextHeap(gc_alloc_context* acontext, int thread_number) = 0;
+    
+    // Returns whether or not this object resides in an ephemeral generation.
+    virtual BOOL IsEphemeral(Object* object) = 0;
+
+    // Blocks until a GC is complete, returning a code indicating the wait was successful.
+    virtual uint32_t WaitUntilGCComplete(BOOL bConsiderGCStart = FALSE) = 0;
+
+    // "Fixes" an allocation context by binding its allocation pointer to a
+    // location on the heap.
+    virtual void FixAllocContext(gc_alloc_context* acontext, BOOL lockp, void* arg, void* heap) = 0;
+
+    // Gets the total survived size plus the total allocated bytes on the heap.
+    virtual size_t GetCurrentObjSize() = 0;
+
+    // Sets whether or not a GC is in progress.
+    virtual void SetGCInProgress(BOOL fInProgress) = 0;
+
+
+    /*
+    ===========================================================================
+    Allocation routines. These all call into the GC's allocator and may trigger a garbage
+    collection.
+    ===========================================================================
+    */
+
+    // Allocates an object on the given allocation context with the given size and flags.
+    virtual Object* Alloc(gc_alloc_context* acontext, size_t size, uint32_t flags) = 0;
+
+    // Allocates an object on the default allocation context with the given size and flags.
+    virtual Object* Alloc(size_t size, uint32_t flags) = 0;
+
+    // Allocates an object on the large object heap with the given size and flags.
+    virtual Object* AllocLHeap(size_t size, uint32_t flags) = 0;
+
+    // Allocates an object on the default allocation context, aligned to 64 bits,
+    // with the given size and flags.
+    virtual Object* AllocAlign8 (size_t size, uint32_t flags) = 0;
+
+    // Allocates an object on the given allocation context, aligned to 64 bits,
+    // with the given size and flags.
+    virtual Object* AllocAlign8 (gc_alloc_context* acontext, size_t size, uint32_t flags) = 0;
+
+    // If allocating on the LOH, blocks if a background GC is occuring in the
+    // background sweep phase.
+    virtual void PublishObject(uint8_t* obj) = 0;
```

Comment:
well, a bgc cannot happen during the bgc sweep phase :) this API is used for large object allocator to coordinate with concurrent marking.